src/ocean/core/Option.d

```diff
@@ -0,0 +1,236 @@
+/******************************************************************************
+
+    Wraps any type in a struct that also contains boolean field indicating if
+    value is in defined state.
+
+    If T is a value type, `Option!(T)` is value type too.
+
+    Copyright:
+        Copyright (c) 2009-2016 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+
+*******************************************************************************/
+
+module ocean.core.Option;
+
+version(UnitTest)
+{
+    import ocean.core.Test;
+}
+
+/// ditto
+struct Option ( T )
+{
+    /// Alias to make code working with undefined state more readable
+    public const undefined = Option!(T).init;
+
+    /// wrapped arbitrary value
+    private T value;
+
+    /// flag indicating if stored value is in defined state
+    private bool defined = false;
+
+    /**************************************************************************
+
+        Puts `this` into defined state
+
+        Params:
+            rhs = value to assign
+
+    **************************************************************************/
+
+    public void opAssign ( T rhs )
+    {
+        this.defined = true;
+        this.value = rhs;
+    }
+
+    /**************************************************************************
+
+        Puts `this` into undefined state
+
+    **************************************************************************/
+
+    public void reset()
+    {
+        this.defined = false;
```

Comment:
Maybe reset `this = this.init` so that `this.value` is reset as well? Otherwise, if `T` is or contains reference types, there could be dangling references (which shouldn't be possible to use, but IMHO it's a good idea to avoid them if possible), or garbage collection could be prevented by keeping a pointer to an object that is in fact not used any more.